src/whir/batching.rs

```diff
@@ -0,0 +1,819 @@
+{
+    if stir_list.is_empty() {
+        return true;
+    }
+
+    let (basic_indexes, mut computed_stir) = stir_list[0].clone();
+    let mut multiplier = *batching_randomness;
+
+    for (multi_path, stir_queries) in &stir_list[1..] {
+        if !multi_path
+            .leaf_indexes
+            .iter()
+            .zip(basic_indexes.leaf_indexes.iter())
+            .all(|(lhs, rhs)| lhs == rhs)
+        {
+            return false;
+        }
+
+        fma_stir_queries(multiplier, stir_queries.as_slice(), &mut computed_stir);
+        multiplier *= batching_randomness;
+    }
+
+    let result = expected_stir_value == computed_stir;
+    result
+}
+
+#[cfg(not(debug_assertions))]
+fn validate_stir_queries<F, MerkleConfig>(
+    _batching_randomness: &F,
+    _expected_stir_value: &[Vec<F>],
+    _stir_list: &[(MultiPath<MerkleConfig>, Vec<Vec<F>>)],
+) -> bool
+where
+    F: Field,
+    MerkleConfig: MTConfig<Leaf = [F]>,
+{
+    true
+}
+
+#[derive(Clone, CanonicalSerialize, CanonicalDeserialize)]
+pub struct BatchedWhirProof<MerkleConfig, F>
+where
+    MerkleConfig: MTConfig<Leaf = [F]>,
+    F: Sized + Clone + CanonicalSerialize + CanonicalDeserialize,
+{
+    ///
+    /// List of Merkle paths and the PCP slot value for each individual Merkle
+    /// tree in the batch. The order of MultiPath must match the order of roots
+    /// in the commitment.
+    ///
+    pub first_round_paths: Vec<(MultiPath<MerkleConfig>, Vec<Vec<F>>)>,
+
+    /// List of Merkle Paths after first batched round
+    pub merkle_paths: Vec<(MultiPath<MerkleConfig>, Vec<Vec<F>>)>,
+
+    /// Statement value at a random point.
+    pub statement_values_at_random_point: Vec<F>,
+}
+
+///
+/// This is the list of [Witness] structs as if each polynomial will be folded
+/// independently. The witnesses are individually committed to the FS-transcript.
+///
+pub struct BatchedWitness<F, MerkleConfig: MTConfig> {
+    pub(crate) witness_list: Vec<Witness<F, MerkleConfig>>,
+    pub(crate) batching_randomness: F,
+}
+
+impl<F, MerkleConfig> BatchedWitness<F, MerkleConfig>
+where
+    MerkleConfig: MTConfig,
+    F: Field,
+{
+    ///
+    /// Combines all polynomials and leaf nodes weighted by powers of
+    /// batching_randomness and returns the pair (batched_polynomial,
+    /// batched_oracle). This function is expensive O(2^num_variables).
+    ///
+    pub fn batch_witnesses(&self) -> (CoefficientList<F>, Vec<F>) {
+        assert!(
+            self.witness_list.len() > 1,
+            "Batched witness list should have at least 2 elements"
+        );
+
+        assert!(self.witness_list.iter().all(|wit| wit
+            .polynomial
+            .coeffs()
+            .len()
+            .is_power_of_two()
+            && wit.merkle_leaves.len().is_power_of_two()
+            && wit.polynomial.num_variables() == self.witness_list[0].polynomial.num_variables()
+            && wit.merkle_leaves.len() == self.witness_list[0].merkle_leaves.len()),);
```

Comment:
It's best to keep `assert` expressions as simple as possible, and definitely break them down to individual clauses instead of using `and`. I learned this the hard way, getting back a crash report from a client that said "assertion failed: clause_1 && clause_2 && clause_3", and had to ship an update first to figure out which clause failed to diagonse the problem)

```rust
for witness in &self.witness_list {
  assert!(witness.coeffs().len().is_power_of_two());
  assert!(witness.merkle_leaves.len().is_power_of_two());
  // ...
}
```